src/cargo/ops/registry/publish.rs

```diff
@@ -57,151 +64,192 @@ pub struct PublishOpts<'gctx> {
 }
 
 pub fn publish(ws: &Workspace<'_>, opts: &PublishOpts<'_>) -> CargoResult<()> {
+    let multi_package_mode = ws.gctx().cli_unstable().package_workspace;
     let specs = opts.to_publish.to_package_id_specs(ws)?;
-    if specs.len() > 1 {
-        bail!("the `-p` argument must be specified to select a single package to publish")
+
+    if !multi_package_mode {
+        if specs.len() > 1 {
+            bail!("the `-p` argument must be specified to select a single package to publish")
+        }
+        if Packages::Default == opts.to_publish && ws.is_virtual() {
+            bail!("the `-p` argument must be specified in the root of a virtual workspace")
+        }
     }
-    if Packages::Default == opts.to_publish && ws.is_virtual() {
-        bail!("the `-p` argument must be specified in the root of a virtual workspace")
+
+    let member_ids: Vec<_> = ws.members().map(|p| p.package_id()).collect();
+    // Check that the specs match members.
+    for spec in &specs {
+        spec.query(member_ids.clone())?;
     }
-    let member_ids = ws.members().map(|p| p.package_id());
-    // Check that the spec matches exactly one member.
-    specs[0].query(member_ids)?;
     let mut pkgs = ws.members_with_features(&specs, &opts.cli_features)?;
     // In `members_with_features_old`, it will add "current" package (determined by the cwd)
     // So we need filter
     pkgs = pkgs
         .into_iter()
         .filter(|(m, _)| specs.iter().any(|spec| spec.matches(m.package_id())))
         .collect();
-    // Double check. It is safe theoretically, unless logic has updated.
-    assert_eq!(pkgs.len(), 1);
 
-    let (pkg, cli_features) = pkgs.pop().unwrap();
-
-    let mut publish_registry = match opts.reg_or_index.as_ref() {
-        Some(RegistryOrIndex::Registry(registry)) => Some(registry.clone()),
-        _ => None,
-    };
-    if let Some(ref allowed_registries) = *pkg.publish() {
-        if publish_registry.is_none() && allowed_registries.len() == 1 {
```

Comment:
nit: a potential refactor that could be broken out is to switch this code into the style of `match` in the new code.